lib/plurimath/xml_engine/oga.rb

```diff
@@ -0,0 +1,231 @@
+            [e.name, e.value]
+          end
+        end
+
+        def locate(xpath)
+          @wrapped.xpath(xpath).map do |i|
+            case i
+            when ::Oga::XML::Text
+              get_text_from(i)
+            when ::Oga::XML::Comment
+              i
+            else
+              Node.new(i)
+            end
+          end
+        end
+
+        def name
+          @wrapped.name
+        end
+
+        def name=(new_name)
+          @wrapped.name = new_name
+        end
+      end
+
+      class Document < Node
+      end
+
+      # Dump the tree just as if we were Ox. This is a limited implementation.
+      class Dumper
+        def initialize(tree, indent: nil)
+          @tree = tree
+          @indent = indent
+          @depth = 0
+          @out = ""
+        end
+
+        def dump(node = @tree)
+          case node
+          when ::Oga::XML::Element
+            if node.children.length == 0
+              indent
+              @out += "<#{node.name}#{dump_attrs(node)}/>"
+              line_break
+            elsif node.children.length == 1 && node.children.first.is_a?(::Oga::XML::Text)
+              indent
+              @out += "<#{node.name}#{dump_attrs(node)}>"
+              @out += entities(node.children.first.text)
+              @out += "</#{node.name}>"
+              line_break
+            else
+              indent
+              @out += "<#{node.name}#{dump_attrs(node)}>"
+              line_break
+              @depth += 1
+              node.children.each { |i| dump(i) }
+              @depth -= 1
+              indent
+              @out += "</#{node.name}>"
+              line_break
+            end
+          when ::Oga::XML::Text
+            indent
+            @out += entities(node.text)
+            line_break
+          else
+            raise NotImplementedError
+          end
+
+          self
+        end
+
+        attr_reader :out
+
+        private
+
+        def dump_attrs(node)
+          node.attributes.map do |i|
+            #if i.namespace
+            #  %{ #{i.namespace.name}:#{i.name}="#{attr_entities i.value}"}
+            #else
+            %{ #{i.name}="#{attr_entities i.value}"}
+            #end
+          end.join
+        end
+
+        ORD_AMP="&".ord
+        ORD_LT="<".ord
+        ORD_GT=">".ord
+        ORD_APOS="'".ord
+        ORD_QUOT='"'.ord
+        ORD_NEWLINE="\n".ord
+        ORD_CARRIAGERETURN="\r".ord
+
+        def entities(text,attr=false)
+          text.to_s.chars.map(&:ord).map do |i|
+            if i == ORD_AMP
+              "&amp;"
+            elsif i == ORD_LT
+              "&lt;"
+            elsif i == ORD_GT
+              "&gt;"
+            elsif i == ORD_APOS && attr
+              "&apos;"
+            elsif i == ORD_QUOT && attr
+              "&quot;"
+            elsif i == ORD_NEWLINE || i == ORD_CARRIAGERETURN
+              i.chr("utf-8")
+            elsif i == ORD_AMP || i == ORD_LT || i == ORD_GT ||
```

Comment:
shouldn't case be used instead of if?

Reply:
There's a check for `&& attr` on line 200 and 202, so unfortunately we can't use `case`.